openstack/utils/choose_version.go

```diff
@@ -109,3 +113,103 @@ func ChooseVersion(client *gophercloud.ProviderClient, recognized []*Version) (*
 
 	return highest, endpoint, nil
 }
+
+type SupportedMicroversions struct {
+	MaxMajor int
+	MaxMinor int
+	MinMajor int
+	MinMinor int
+}
+
+// GetSupportedMicroversions returns the minimum and maximum microversion that is supported by the ServiceClient Endpoint.
+func GetSupportedMicroversions(client *gophercloud.ServiceClient) (SupportedMicroversions, error) {
+	type valueResp struct {
+		ID         string `json:"id"`
+		Status     string `json:"status"`
+		Version    string `json:"version"`
+		MinVersion string `json:"min_version"`
+	}
+
+	type response struct {
+		Version valueResp `json:"version"`
+	}
+	var supportedMicroversions SupportedMicroversions
+	var resp response
+	_, err := client.Get(client.Endpoint, &resp, &gophercloud.RequestOpts{
+		OkCodes: []int{200, 300},
+	})
+
+	if err != nil {
+		return supportedMicroversions, err
+	}
+
+	supportedMicroversions.MinMajor, supportedMicroversions.MinMinor, err = ParseMicroversion(resp.Version.MinVersion)
+	if err != nil {
+		return supportedMicroversions, err
+	}
+
+	supportedMicroversions.MaxMajor, supportedMicroversions.MaxMinor, err = ParseMicroversion(resp.Version.Version)
+	if err != nil {
+		return supportedMicroversions, err
+	}
+
+	return supportedMicroversions, nil
+}
+
+// RequireMicroversion checks that the required microversion is supported and
+// returns a ServiceClient with the microversion set.
+func RequireMicroversion(client gophercloud.ServiceClient, required string) (gophercloud.ServiceClient, error) {
+	supportedMicroversions, err := GetSupportedMicroversions(&client)
+	if err != nil {
+		return client, fmt.Errorf("unable to determine supported microversions: %w", err)
+	}
+	supported, err := MicroversionSupported(required, supportedMicroversions)
+	if err != nil {
+		return client, err
+	}
+	if !supported {
+		return client, fmt.Errorf("microversion %s not supported. Supported versions: %v", required, supportedMicroversions)
+	}
+	client.Microversion = required
+	return client, nil
+}
+
+// MicroversionSupported checks if a microversion falls in the supported interval.
+// It returns true if the version is within the interval and false otherwise.
+func MicroversionSupported(version string, supportedMicroversions SupportedMicroversions) (bool, error) {
```

Comment:
Why not make this a method on SupportedMicroversions?

Reply:
Done (and renamed to `IsSupported`)